### ISSUE TITLE
Do not automatically place 'Root' users into every group

### ISSUE THREAD

**Issue by mark0263** [glFusion/glfusion#441]
By default, Root members are also members of every other group created.  This is done by placing all the members of root into the other groups.

Modify so Root members are not automatically placed in new groups.

**Comment by mark0263** [comment]
Couple of thoughts on how to accomplish this...

1. Modify the code so it no longer inserts root members upon save.  The downside to this is that there are several other areas of the code that could be impacted, since the assumption has always been, root users have access and are members of all other groups.  

2. Add an attribute to a group to exclude root members - it may be as simple as a form field so the save does not automatically populate the membership with group members.  

Thoughts?

**Comment by leegarner** [comment]
#2 would work, but my thought is to go with #1. For backward compatibility functions like SEC_hasTopicAccess() and maybe even SEC_inGroup() could be modified to return true for Root users, but I think SEC_hasRights() should be leveraged as much as possible for access to site features and admin functions.  A quick grep for SEC_inGroup through private turned up very few instances where the check was not for Root and in at least almost all cases excluding Root and including only directly-included users appears to be the intent.

**Comment by mark0263** [comment]
I like the idea of going the #2 routes as well.  Need some thoughts on how rights and groups align.  Currently, most areas have both rights with a corresponding group (i.e.; Groups Admin has groups.edit and groups.delete rights).  Do we need to maintain the 1:1 ratio of groups to rights?  Another way to put it - Admin Actions currently has no groups or rights associated - if I create a right adminactions.admin - do I need to create a group called Admin Actions Admin - I don't think we do - I think we can assign the rights to Root and then the site owner can assign it to any other group they see fit.  

Thoughts on approach?

**Comment by mark0263** [comment]
Looking at the current code - I found that menu.admin is done with just a right and no group.

**Comment by leegarner** [comment]
That's exactly what I was thinking. I've been removing admin groups from my
plugins and just assigning the plugin.admin right to the Root group. Since
the right can be taken away from Root, I'm thinking maybe SEC_hasRights()
should always return true if SEC_inGroup('Root'); In this case plugins
don't even need to create their own groups, but admins can still do so if
they like.

**Comment by mark0263** [comment]
Since we implemented the system.root permission - it makes it a bit more difficult to call SEC_inGroup('Root') from SEC_hasRights() - creates an infinite loop :). I agree - I like not having so many groups and think we just go with the features / rights.  I'll fix the infinite loop issue too...

**Comment by mark0263** [comment]
Easy fix on the infinite loop.  The way it is coded now - if you are a member of Root group, or have the system.root feature assigned - that user will have every feature / right and group access, without actually being a member of another group. 

I want to update the Group edit screen so you cannot de-select a feature, since it would not actually do anything...

Probably need to test this pretty well to ensure it has the desired affect...

**Comment by leegarner** [comment]
Sounds perfect, Not sure about blocking the de-selecting a feature though, wouldn't that prevent admins from mixing and matching permissions? For Root I see limiting edit access, don't want to make it too easy for the admins to lock themselves out. Ultimately as the groups become unnecessary glFusion could stop including them, but not remove them during upgrades. Just a note in the readme that the groups can be removed for easier administration. Most sites probably don't even use them for admin access control at all if they only have one or two admins.

**Comment by mark0263** [comment]
For blocking the de-selecting it would be for the Root group only - all others you can mix / match as needed.

Right now the way it is all coded - using autotag.admin as an example, I can de-select it from the Root features, but it really doesn't matter since SEC_hasRights() now returns true if you are in Root - my thought was - for the Root group only - remove the option to de-select a feature - since it really doesn't remove it from root.

**Comment by leegarner** [comment]
Perfect, that's what I was thinking.

**Comment by mark0263** [comment]
I've committed all of the changes that should do the following:
 - creating a new group - no assignments are made into the group_assignments table to force 'root' membership (this allows creating groups that do not automatically include root members
 - essentially disabled editing the Root group - it will show all groups / rights assigned and cannot be modified
 - SEC_inGroup() will return true if user is in Root or has system.root permissions on any group check
 - SEC_hasRights() will return true if user is in Root or has system.root permissions on any feature / right

**Comment by mark0263** [comment]
Ad I'm looking at the group_assignments table - it seems silly to have an entry for every user where they are assigned to the Logged-in Users group and the all Users group - these can easily be added to everyone in lib-common:

```
if ( !COM_isAnonUser() ) {
    $_GROUPS = SEC_getUserGroups( $_USER['uid'] );
    $_GROUPS['All Users'] = 2;
    $_GROUPS['Logged-in Users'] = 13;
} else {
    $_GROUPS = SEC_getUserGroups( 1 );
    $_GROUPS['All Users'] = 2;
}
```

This would require a small change to the Email Group functionality to simply to a select * on the users table instead of the group_assignment table.  Other than this functionality, I don't see any other area that is leveraging these groups.

Thoughts on removing the group assignments and just handling this in the code?

**Comment by walterrowe** [comment]
Seems reasonable to me.

**Comment by leegarner** [comment]
Agreed.

**Comment by leegarner** [comment]
A thought occurs to me... as we're leaning more on Features and less on Groups for access control to site functions (as opposed to content), we should have a method of editing features. I just needed to verify what group I had given a feature to and had to look in the gl_access table. I started a Feature class some time ago for this, unless you (Mark) have started working on a feature-editing function I'll revive it and create a PR to add it into C&C.
